eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+package org.axonframework.eventsourcing;
+
+import org.axonframework.common.Assert;
+import org.axonframework.common.stream.BlockingStream;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.eventhandling.GenericTrackedEventMessage;
+import org.axonframework.eventhandling.GenericTrackedSourcedEventMessage;
+import org.axonframework.eventhandling.MultiSourceTrackingToken;
+import org.axonframework.eventhandling.TrackedEventMessage;
+import org.axonframework.eventhandling.TrackedSourcedEventMessage;
+import org.axonframework.eventhandling.TrackingToken;
+import org.axonframework.messaging.StreamableMessageSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Implementation which allows for tracking processors to process messages from an arbitrary number of sources. The
+ * order in which messages from each stream are consumed is configurable but defaults to the oldest message available
+ * (using the event's timestamp). When the stream is polled for a specified duration, each stream is called with
+ * {@link MultiSourceMessageStream#hasNextAvailable()} except for the last stream configured by the
+ * {@link MultiStreamableMessageSource.Builder#addMessageSource(String, StreamableMessageSource)} or by explicit
+ * configuration using {@link MultiStreamableMessageSource.Builder#configureLongPollingSource(String)}. This stream long
+ * polls for a fraction of the specified duration before looping through the sources again repeating until the duration
+ * has been met. This ensure the highest chance of a consumable message being found.
+ *
+ * @author Greg Woods
+ * @since 4.x
+ */
+public class MultiStreamableMessageSource implements StreamableMessageSource<TrackedEventMessage<?>> {
+
+    private static final Logger logger = LoggerFactory.getLogger(MultiStreamableMessageSource.class);
+
+    private final Map<String, StreamableMessageSource<TrackedEventMessage<?>>> eventStreams;
+    private MultiSourceTrackingToken trackingToken;
+    private final Comparator<TrackedSourcedEventMessage<?>> comparator;
```

Comment:
I personally find `comparator` slightly generic. I'd opt for a rename to `trackedEventMessageComparator` or `trackedEventComparator` for example.

Reply:
This will obviously also rename the builder method and builder private field to align with this rename.